spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java

```diff
@@ -0,0 +1,385 @@
+          } else {
+            return greaterThan(
+                unquote(predicate.children()[SECOND_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value()));
+          }
+
+        case LT_EQ:
+          if (predicate.children()[SECOND_ORDINAL] instanceof LiteralValue) {
+            return lessThanOrEqual(
+                unquote(predicate.children()[FIRST_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[1]).value()));
+          } else {
+            return greaterThanOrEqual(
+                unquote(predicate.children()[SECOND_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value()));
+          }
+
+        case GT:
+          if (predicate.children()[SECOND_ORDINAL] instanceof LiteralValue) {
+            return greaterThan(
+                unquote(predicate.children()[FIRST_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[SECOND_ORDINAL]).value()));
+          } else {
+            return lessThan(
+                unquote(predicate.children()[SECOND_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value()));
+          }
+
+        case GT_EQ:
+          if (predicate.children()[SECOND_ORDINAL] instanceof LiteralValue) {
+            return greaterThanOrEqual(
+                unquote(predicate.children()[FIRST_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[1]).value()));
+          } else {
+            return lessThanOrEqual(
+                unquote(predicate.children()[SECOND_ORDINAL].toString()),
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value()));
+          }
+
+        case EQ: // used for both eq and null-safe-eq
+          Object value;
+          String attributeName;
+          if (predicate.children()[SECOND_ORDINAL] instanceof LiteralValue) {
+            attributeName = predicate.children()[FIRST_ORDINAL].toString();
+            value = convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[1]).value());
+          } else {
+            attributeName = predicate.children()[SECOND_ORDINAL].toString();
+            value =
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value());
+          }
+
+          if (predicate.name().equals("=")) {
+            // comparison with null in normal equality is always null. this is probably a mistake.
+            Preconditions.checkNotNull(
+                value, "Expression is always false (eq is not null-safe): %s", predicate);
```

Comment:
Question: will this fail the query?

So if somebody writes `select a, b from foo where c = null`, the query will throw? Would we be better off logging a warning and then using `Expressions.alwaysFalse()` (or null), or is throwing more user friendly since probably they didn’t mean for that to be the query — the query basically will not match anything right?

Genuinely asking as I’m not sure if people use this pattern at all. You’re right it seems like a mistake.

Maybe we should add a unit test that this throws the specified exception?

I’m genuinely not sure which is best so just asking. I’m good with it as is.

Reply:
This is actually borrowed from `SparkFilters`

Reply:
Ah then absolutely keep it :)

Reply:
It is unexpected and I think we wrap the conversion logic into try-catch, so we should be fine.